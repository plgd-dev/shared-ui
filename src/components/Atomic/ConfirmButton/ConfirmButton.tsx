import { FC, useState } from 'react'
import ConfirmModal from '../ConfirmModal'
import Button from '../Button'
import { defaultProps, Props } from './ConfirmButton.types'

const ConfirmButton: FC<Props> = (props) => {
    const { onConfirm, children, confirmButtonText, cancelButtonText, title, body, loading, closeOnConfirm, modalProps, ...rest } = {
        ...defaultProps,
        ...props,
    }
    const [show, setShow] = useState(false)

    const onConfirmClick = (onClose: () => void) => {
        onConfirm(onClose)

        if (closeOnConfirm) {
            setShow(false)
        }
    }

    return (
        <>
            <Button {...rest} loading={loading} onClick={() => setShow(true)}>
                {children}
            </Button>
            <ConfirmModal
                {...modalProps}
                body={body}
                cancelButtonText={cancelButtonText}
                confirmButtonText={confirmButtonText}
                loading={loading}
                onClose={() => setShow(false)}
                onConfirm={onConfirmClick}
                show={show}
                title={title}
            />
        </>
    )
}

ConfirmButton.displayName = 'ConfirmButton'

export default ConfirmButton
