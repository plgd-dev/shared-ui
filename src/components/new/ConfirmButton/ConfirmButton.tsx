import { FC, useState } from 'react'
import ConfirmModal from '../ConfirmModal'
import Button from '../Button'
import { defaultProps, Props } from './ConfirmButton.types'

const ConfirmButton: FC<Props> = (props) => {
    const { onConfirm, children, confirmButtonText, cancelButtonText, title, body, loading, closeOnConfirm, modalProps, ...rest } = props
    const [show, setShow] = useState(false)

    const onConfirmClick = (onClose: () => void) => {
        onConfirm(onClose)

        if (closeOnConfirm) {
            setShow(false)
        }
    }

    return (
        <>
            <Button {...rest} loading={loading} onClick={(e) => setShow(true)}>
                {children}
            </Button>
            <ConfirmModal
                {...modalProps}
                show={show}
                onClose={() => setShow(false)}
                title={title}
                body={body}
                loading={loading}
                cancelButtonText={cancelButtonText}
                confirmButtonText={confirmButtonText}
                onConfirm={onConfirmClick}
            />
        </>
    )
}

ConfirmButton.displayName = 'ConfirmButton'
ConfirmButton.defaultProps = defaultProps

export default ConfirmButton
