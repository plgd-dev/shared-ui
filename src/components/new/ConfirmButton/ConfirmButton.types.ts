import { ReactNode } from 'react'

export type Props = {
    body: ReactNode
    cancelButtonText: string
    children: ReactNode
    closeOnConfirm?: boolean
    confirmButtonText: string
    loading?: boolean
    modalProps?: object
    onConfirm: (onClose: () => void) => void
    title: string
}

export const defaultProps = {
    confirmButtonText: '',
    cancelButtonText: '',
    closeOnConfirm: true,
    modalProps: {},
}
