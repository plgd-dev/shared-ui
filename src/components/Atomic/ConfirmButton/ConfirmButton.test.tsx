import { render, fireEvent, waitFor } from '@testing-library/react'
import ConfirmButton from './ConfirmButton'

describe('<ConfirmButton>', () => {
    it('renders without crashing', () => {
        const { container, asFragment } = render(
            <ConfirmButton body={<div>body</div>} cancelButtonText='Cancel' confirmButtonText='Confirm' onConfirm={() => {}} title='Title'>
                Action
            </ConfirmButton>
        )
        expect(container).toBeInTheDocument()

        expect(asFragment()).toMatchSnapshot()
    })

    it('opens ConfirmModal when clicked', () => {
        const { getByRole, queryByRole } = render(
            <ConfirmButton body={<div>body</div>} cancelButtonText='Cancel' confirmButtonText='Confirm' onConfirm={() => {}} title='Title'>
                Action
            </ConfirmButton>
        )
        const button = getByRole('button')
        fireEvent.click(button)
        const modal = queryByRole('dialog')
        expect(modal).toBeInTheDocument()
    })

    it('calls onConfirm when confirm button in modal is clicked', () => {
        const mockOnConfirm = jest.fn()
        const { getByRole, getByText } = render(
            <ConfirmButton body={<div>body</div>} cancelButtonText='Cancel' confirmButtonText='Confirm' onConfirm={mockOnConfirm} title='Title'>
                Action
            </ConfirmButton>
        )
        const button = getByRole('button')
        fireEvent.click(button)
        const confirmButton = getByText('Confirm')
        fireEvent.click(confirmButton)
        expect(mockOnConfirm).toHaveBeenCalled()
    })

    it('closes ConfirmModal when closeOnConfirm is true', async () => {
        const { getByRole, queryByRole } = render(
            <ConfirmButton closeOnConfirm body={<div>body</div>} cancelButtonText='Cancel' confirmButtonText='Confirm' onConfirm={() => {}} title='Title'>
                Action
            </ConfirmButton>
        )
        const button = getByRole('button')
        fireEvent.click(button)
        const confirmButton = getByRole('button', { name: /Confirm/i })
        fireEvent.click(confirmButton)
        const modal = queryByRole('dialog')

        await waitFor(() => {
            expect(modal).not.toBeInTheDocument()
        })
    })

    it('does not close ConfirmModal when closeOnConfirm is false', () => {
        const { getByRole, queryByRole } = render(
            <ConfirmButton
                body={<div>body</div>}
                cancelButtonText='Cancel'
                closeOnConfirm={false}
                confirmButtonText='Confirm'
                onConfirm={() => {}}
                title='Title'
            >
                Action
            </ConfirmButton>
        )
        const button = getByRole('button')
        fireEvent.click(button)
        const confirmButton = getByRole('button', { name: /Confirm/i })
        fireEvent.click(confirmButton)
        const modal = queryByRole('dialog')
        expect(modal).toBeInTheDocument()
    })
})
