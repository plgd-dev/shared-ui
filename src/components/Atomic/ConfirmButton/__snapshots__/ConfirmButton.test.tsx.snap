// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfirmButton> renders without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 600;
  text-align: center;
  border-radius: 8px;
  white-space: nowrap;
  box-sizing: border-box;
  cursor: pointer;
  border: 1px solid;
  padding: 10px 24px;
  font-size: 14px;
  line-height: 16px;
  height: 48px;
}

<button
    class="emotion-0"
    tabindex="1"
    type="submit"
  >
    Action
  </button>
</DocumentFragment>
`;
